pkg/compactor/visit_marker.go

```diff
@@ -0,0 +1,220 @@
+	ticker := time.NewTicker(visitMarkerFileUpdateInterval)
+	defer ticker.Stop()
+heartBeat:
+	for {
+		v.MarkInProgress(ctx)
+
+		select {
+		case <-ctx.Done():
+			level.Warn(v.getLogger()).Log("msg", "visit marker heart beat got cancelled")
+			v.MarkPending(context.Background())
+			break heartBeat
+		case <-ticker.C:
+			continue
+		case err := <-errChan:
+			if err == nil {
+				level.Info(v.getLogger()).Log("msg", "update visit marker to completed status")
+				v.MarkCompleted(ctx)
+			} else {
+				level.Warn(v.getLogger()).Log("msg", "stop visit marker heart beat due to error", "err", err)
+				if compact.IsHaltError(err) {
+					level.Info(v.getLogger()).Log("msg", "update visit marker to failed status", "err", err)
+					v.MarkFailed(ctx)
+				} else {
+					level.Info(v.getLogger()).Log("msg", "update visit marker to pending status", "err", err)
+					v.MarkPending(ctx)
+				}
+			}
+			break heartBeat
+		}
+	}
+	level.Info(v.getLogger()).Log("msg", "stop visit marker heart beat")
+	if deleteOnExit {
+		level.Info(v.getLogger()).Log("msg", "delete visit marker when exiting heart beat")
+		v.DeleteVisitMarker(context.Background())
+	}
+}
+
+func (v *VisitMarkerManager) MarkInProgress(ctx context.Context) {
+	v.visitMarker.MarkInProgress(v.ownerIdentifier)
+	if err := v.updateVisitMarker(ctx); err != nil {
+		level.Error(v.getLogger()).Log("msg", "unable to upsert visit marker file content", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "marked in progress")
+}
+
+func (v *VisitMarkerManager) MarkPending(ctx context.Context) {
+	v.visitMarker.MarkPending(v.ownerIdentifier)
+	if err := v.updateVisitMarker(ctx); err != nil {
+		level.Error(v.getLogger()).Log("msg", "unable to upsert visit marker file content", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "marked pending")
+}
+
+func (v *VisitMarkerManager) MarkCompleted(ctx context.Context) {
+	v.visitMarker.MarkCompleted(v.ownerIdentifier)
+	if err := v.updateVisitMarker(ctx); err != nil {
+		level.Error(v.getLogger()).Log("msg", "unable to upsert visit marker file content", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "marked completed")
+}
+
+func (v *VisitMarkerManager) MarkFailed(ctx context.Context) {
+	v.visitMarker.MarkFailed(v.ownerIdentifier)
+	if err := v.updateVisitMarker(ctx); err != nil {
+		level.Error(v.getLogger()).Log("msg", "unable to upsert visit marker file content", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "marked failed")
+}
+
+func (v *VisitMarkerManager) DeleteVisitMarker(ctx context.Context) {
+	if err := v.bkt.Delete(ctx, v.visitMarker.GetVisitMarkerFilePath()); err != nil {
+		level.Error(v.getLogger()).Log("msg", "failed to delete visit marker", "err", err)
+		return
+	}
+	level.Debug(v.getLogger()).Log("msg", "visit marker deleted")
+}
+
+func (v *VisitMarkerManager) ReloadVisitMarker(ctx context.Context) error {
```

Comment:
Where is this being used?

Reply:
It is not used. Added this function as a helper in case this is needed in future. I am ok to remove it for now.